### ISSUE TITLE
Support narrowing `dict`s on `.get` calls

### ISSUE THREAD

**Issue by MeGaGiGaGon** [astral-sh/ty#3394]
### Summary

It would be nice if ty supported this pattern:
https://play.ty.dev/9df66eee-d724-40d5-ba13-311c2bf78b62
```py
from typing import reveal_type

def bar(data: dict[str, int | None]):
    if data["value"] is not None:
        reveal_type(data["value"])  # Revealed type: `int` 

    if data.get("value") is not None:
        # Expected: `int`
        reveal_type(data["value"])  # Revealed type: `int | None`
```
(For reference, currently only pyrefly supports it (multiplay e7b26a81fcd8170282d4ed2cb1612ba7))
(In my actual code I ran into this in, it is instead a negative pattern like `if data.get("value") is None: continue`)

### Version

ty 0.0.33 (c512d8425 2026-04-28)

**Comment by sharkdp** [comment]
Is there really a need to add this? I feel like this migh encourage a pattern of code that feels non-ideal? Why would I access the dictionary twice, once via `.get` and then via subscription, when I can store the value instead?
```py
value = data.get("value")
if value is not None:
    reveal_type(value)  # `str`
```

**Comment by MeGaGiGaGon** [comment]
Without `.get`:
```py
if "value" not in data:
    continue
if data["value"] is None:
    continue
operation(data["value"])
```
With this narrowing:
```py
if data.get("value") is None:
    continue
operation(data["value"])
```
Your idea:
```py
value = data.get("value")
if value is None:
    continue
operation(value)
```
The part that I like most about this narrowing is how it is the conceptually easiest out of them, at least for me in the code I was writing. It's a one statement guard that is easy to keep track of, since the mental note is "`data["value"]` is in a good state". Both without `.get` and your idea have two statements to combine into one idea mentally to get to the same mental note. Your way is probably more idiomatic though.

**Comment by AlexWaygood** [comment]
I also feel like this pattern is pretty unusual, and not what I'd really consider idiomatic. @charliermarsh's PR seems to validate this: there is [extremely little](https://github.com/astral-sh/ruff/pull/24965#issuecomment-4362083620) ecosystem impact here.

I'm quite surprised that pyrefly added support for this; no other type checker does. There's an infinite universe of narrowing patterns that we _could_ support, but each requires its own special casing. We have to draw the line somewhere; to me, this clearly falls on the other side of the line.

**Comment by charliermarsh** [comment]
It's empirically an unpopular opinion in the thread but... does it matter if it's idiomatic? It's correct and leads to better, more informed results. Is there a material downside to supporting it?

**Comment by MeGaGiGaGon** [comment]
Re "why did pyrefly add this", I believe they have a more general version than what @charliermarsh's pr currently implements (relevant PR: https://github.com/facebook/pyrefly/pull/1390), for example like this:
[sandbox](https://pyrefly.org/sandbox/?project=N4IgZglgNgpgziAXKOBDAdgEwEYHsAeAdAA4CeSImMYABGLrgBSaoAuqiNmEAxqwNpxWAJwA0NCOlY0APjSHCAugEpEAHXQ0tE2hDiShGHjGZtUhAOYxWjNSABuqKAFcYd5eMmtVG7X5rCMPYwTgD6rKTEJizslta2Dk6u7spaAMRagcFOMJg0EVGcXiCiIGSBYFCkhKy4ALZQFBkACqQVVfIYOAQ0PLjokBbOwmwQ-YQaGQDKMDA0ABasrMRwiAD0a%2BXUVYS4whZrMOhrmLg8cGt9AxBDI6xjx3R7NKiO0KjYsL39g8Oj-TRcMR7v04BN0GRWPN%2BgBaYLCfQAgC8NDsAGZCABGABMdg0-BgwmEezgig0znQEDqxD2rFyMO4gT4EGCNBRAHIbug9jB2RpuawYYEAI7OCCBTAwgDWMFIMNQPGMcDgbJo7IA7qhhOg%2BegQABfUoK%2B7BABi0BgFDQWDwRDIBqAA)
```py
def foo(data: dict[str, int | str]):
    if isinstance(data.get("value"), int):
        reveal_type(data.get("value"))  #  revealed type: int
```

**Comment by AlexWaygood** [comment]
Should we also add support for `dict.setdefault(...)` narrowing? What about custom classes that have `.get(...)` methods that work the same way?

> Is there a material downside to supporting it?

The material downside is more complexity for us to maintain, and more complexity for our users to remember. Since we _cannot_ support the whole universe of _possible_ type-narrowing guards, at some point we'll probably want to have a comprehensive list in our documentation of all the type-narrowing operations we support. (This is what pyright does.) It's far easier for our users to digest that if it's a short, intuitive list rather than a huge list of entries that's hard to remember and is neither exhaustive nor particularly curated.

**Comment by ratijas** [comment]
I can implement a dict that returns a different type of object on every `.get()` call, and you can't stop me!

```py
class evildict(dict):
    def __init__(self):
        super().__init__()
        self.counter = 0

    def __getitem__(self, key):
        self.counter += 1

        class EvilItem:
            def __init__(self, c: int):
                self.c = c

            def __repr__(self):
                return f"EvilItem({self.c})"

        return EvilItem(self.counter)


d = evildict()
it1 = d[123]
it2 = d[123]
print(it1 == it2)   # False
print(type(it1))    # <class '__main__.evildict.__getitem__.<locals>.EvilItem'>
print(type(it2))    # <class '__main__.evildict.__getitem__.<locals>.EvilItem'>
print(type(it1) == type(it2))  # False
```

Why do you believe properties and especially method calls must result in the same values over and over again?  Just store them in a controlled local variable -- that's what the whole `:=` walrus operator was added for!

**Comment by ratijas** [comment]
Technically, `isinstance` can also be tricked, through metaclasses, to return True/False at random, but I guess such extreme abuses of type system are out of scope for type checker, since it's the user (developer) who deliberately shots themselves in the foot. Meanwhile, my example with an inner class above can be written off as a simple mistake -- after all non-scripting languages usually won't create a new _class object_, and I've definitely seen such ad-hoc classes in practice.

**Comment by sharkdp** [comment]
Given that this is also unsound, I'm going to make a call here and close this as "not planned". If someone feels strongly that this is a mistake, please feel free to comment and we can reconsider.

**Comment by ibraheemdev** [comment]
> Given that this is also unsound

Our current subscript narrowing is also unsound, so I'm not sure that is a very strong argument against this. That said, I think I agree with @AlexWaygood here in that our current narrowing has relatively well-defined lines that would blurry (so I agree with the decision to close 😄).

**Comment by charliermarsh** [comment]
I am still in favor of this but not strongly and totally fine to be overruled :)

**Comment by MeGaGiGaGon** [comment]
I also still think that this would be nice, but I understand the merit in not doing it too, so C'est la vie

**Comment by AlexWaygood** [comment]
As was noted by @ratijas — if you don't like @sharkdp's suggestion in https://github.com/astral-sh/ty/issues/3394#issuecomment-4398641988, using a walrus here is still more efficient and more concise than repeated `.get()` calls. I believe we should already support narrowing on this pattern:

```py
if (value := data.get("value")) is not None:
    reveal_type(value)  # `str`
```
